src/main/java/iterator/RichIterator.java

```diff
@@ -66,15 +106,23 @@ default Optional<A> nextOptional() {
      * @param f the Consumer
      */
     default void foreach(Consumer<? super A> f) {
-        throw new NotImplementedException();
+        while (hasNext()) {
+            f.accept(next());
+        }
     }
 
     /**
      * @param elem checks if the iterator has the element
      * @return true if exists, false otherwise
      */
     default boolean contains(A elem) {
-        throw new NotImplementedException();
+        while (hasNext()) {
+            if (next().equals(elem)) {
```

Comment:
The tests does not contain `null` checks (maybe I should add some).
But in case the `next` function returns a `null` value this will fail with `NullPointerException`.
A better practice is to use 
```suggestion
            if (Objects.equals(next(), elem)) {
```
Which does basically the same but safe for null values.